src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs

```diff
@@ -701,6 +808,95 @@ public virtual int ReadByte()
+        /// Thrown when the end of the stream is reached before reading <paramref name="count"/> number of bytes.
+        /// </exception>
+        public void ReadExactly(byte[] buffer, int offset, int count)
+        {
+            ValidateBufferArguments(buffer, offset, count);
+
+            _ = ReadAtLeastCore(buffer.AsSpan(offset, count), count, throwOnEndOfStream: true);
+        }
+
+        /// <summary>
+        /// Reads at least a minimum number of bytes from the current stream and advances the position within the stream by the number of bytes read.
+        /// </summary>
+        /// <param name="buffer">A region of memory. When this method returns, the contents of this region are replaced by the bytes read from the current stream.</param>
+        /// <param name="minimumBytes">The minimum number of bytes to read into the buffer.</param>
+        /// <param name="throwOnEndOfStream">
+        /// <see langword="true"/> to throw an exception if the end of the stream is reached before reading <paramref name="minimumBytes"/> of bytes;
+        /// <see langword="false"/> to return less than <paramref name="minimumBytes"/> when the end of the stream is reached.
+        /// The default is <see langword="true"/>.
+        /// </param>
+        /// <returns>
+        /// The total number of bytes read into the buffer. This is guaranteed to be greater than or equal to <paramref name="minimumBytes"/>
+        /// when <paramref name="throwOnEndOfStream"/> is <see langword="true"/>. This will be less than <paramref name="minimumBytes"/> when the
+        /// end of the stream is reached and <paramref name="throwOnEndOfStream"/> is <see langword="false"/>. This can be less than the number
+        /// of bytes allocated in the buffer if that many bytes are not currently available.
+        /// </returns>
+        /// <exception cref="ArgumentOutOfRangeException">
+        /// <paramref name="minimumBytes"/> is negative, or greater than the length of <paramref name="buffer"/>.
+        /// </exception>
+        /// <exception cref="EndOfStreamException">
+        /// Thrown when <paramref name="throwOnEndOfStream"/> is <see langword="true"/> and the end of the stream is reached before reading
+        /// <paramref name="minimumBytes"/> of data.
+        /// </exception>
+        public int ReadAtLeast(Span<byte> buffer, int minimumBytes, bool throwOnEndOfStream = true)
+        {
+            ValidateReadAtLeastArguments(buffer.Length, minimumBytes);
+
+            return ReadAtLeastCore(buffer, minimumBytes, throwOnEndOfStream);
+        }
+
+        // No argument checking is done here. It is up to the caller.
+        private int ReadAtLeastCore(Span<byte> buffer, int minimumBytes, bool throwOnEndOfStream)
+        {
+            Debug.Assert(minimumBytes <= buffer.Length);
+
+            int totalRead = 0;
+            do
+            {
+                int read = Read(buffer.Slice(totalRead));
```

Comment:
The default implementation of `Stream.Read(Span<byte>)` is expensive, as it rents an array from array pool, reads the data into it using the `Read(array)`, copies the content to the span and returns the array to the pool:

https://github.com/dotnet/runtime/blob/65067052e433eda400c5e7cc9f7b21c84640f901/src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs#L677-L695

In BCL all Streams override this method and provide optimal implementation, but it's not guaranteed for other streams.

@eerhardt @stephentoub I am not suggesting that we should have two implementations (one for methods that accept arrays, another for spans), but I would like to hear your opinion on that. Should we care about streams that don't override `Read(Span)` and `Write(Span)`?

Reply:
We use the span/memory-based overloads throughout all of the core library routines that read/write from/to streams.  If streams don't override these methods, they're already going to be deficient compared to what they could be; I don't think adding one more case here significantly moves the needle.

There were a few folks (not including me) in the API review discussion who felt it was important for usability that we expose a byte[]-based overload, but I don't think that means we need to duplicate our code to accommodate it.  If we didn't expose a byte[]-based overload, all use would end up going through the span-based path anyway.